common_benchmark_suite/openxla/benchmark/models/jax/generate_model_artifacts.py

```diff
@@ -0,0 +1,181 @@
+# Copyright 2023 The OpenXLA Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import argparse
+import importlib
+import jax
+import jax.numpy as jnp
+import os
+import pathlib
+import re
+import multiprocessing
+import shutil
+import subprocess
+import sys
+import tarfile
+
+from typing import Any, Optional
+
+# Add openxla dir to the search path.
+sys.path.insert(0, str(pathlib.Path(__file__).parents[4]))
+
+from openxla.benchmark import def_types
+from openxla.benchmark.comparative_suite.jax import model_definitions
+from openxla.benchmark.models.jax import model_interfaces
+
+HLO_FILENAME_REGEX = r".*jit_forward.before_optimizations.txt"
+HLO_STATIC_FILENAME = "xla_hlo_before_optimizations.txt"
+
+
+def _generate_and_save_inputs(model_obj: Any,
+                              model_dir: str) -> tuple[Any, ...]:
+  raw_inputs = model_obj.generate_default_inputs()
+  inputs = model_obj.preprocess(raw_inputs)
+
+  # Save inputs.
+  inputs_dir = os.path.join(model_dir, "inputs")
+  os.makedirs(inputs_dir, exist_ok=True)
+  for idx, input in enumerate(inputs):
+    input_path = os.path.join(inputs_dir, f"input_{idx}.npy")
+    print(f"Saving input {jnp.shape(input)} to {input_path}")
+    jnp.save(input_path, input)
+
+  with tarfile.open(os.path.join(model_dir, "inputs_npy.tgz"), "w:gz") as tar:
```

Comment:
When writing https://github.com/openxla/openxla-benchmark/issues/12#issuecomment-1613839343, I realized we might have multiple raw inputs to be preprocessed for a model. So `model_name/inputs_npy.tgz` isn't enough because it only allows a single preprocessed input.

So we probably at least need to give each raw data a name and store preprocessed input under `model_name/raw_data_name/inputs.npy` and `model_name/raw_data_name/outputs.npy`. Since there is only simple text data now, maybe `model_name/simple_text/inputs.npy` ?

https://github.com/openxla/openxla-benchmark/issues/12#issuecomment-1613839343 proposed a more automated design by adding and using `raw_data_url` and `model` fields in `ModelTestData` to fetch and preprocess the data. But as you mentioned, since the preprocessed data is also part of `ModelTestData`, we run into a chicken or the egg problem, which I think is fine because you can leave the preprocessed data in `ModelTestData` empty first, preprocess and upload the data, then fill out the artifact urls in `ModelTestData`.

Either way seems to be fine for now and has its pros and cons. Also it shouldn't be too hard to revisit this process later.

Reply:
Thanks for the write-up. I've added my thoughts. Let's stick with what we have for now and refactor according the discussions in https://github.com/openxla/openxla-benchmark/issues/12#issuecomment-1613839343.

Reply:
SG. We can iterate on this.